### ISSUE TITLE
Backtest Open/Close price (profits) bug

### ISSUE THREAD

**Issue by creslinux** [freqtrade/freqtrade#961]
From reconciling test results I suspect there is a bug in how backtest calculates its bid and sell price. 
And thus profits per trade. 

TL;DR using close price of current bar when the close of prior bar/candle meets buy/sell logic. 
Should be using open price of current bar when close of prior bar/candle triggers the buy/sell

In this example, 10% drift in 1 trade. 

By way of example. 
I have a strategy that buys on three lime or green on differing MAs
and sells on any of those MAs being Red or Maroon. A simple ribbon indicator. 

<img width="433" alt="screen shot 2018-06-27 at 10 03 05 am" src="https://user-images.githubusercontent.com/34645187/41967522-53661a7c-79f1-11e8-8513-284a962795a0.png">

The trade is exported as having made 0.31 BTC profit on the dates shown. 
`

BTC/USDT   0.31025334          2018-02-12 04:00:00                     2018-02-21 04:00:00

`

In backtest we only know the MAs values when a bar is closed.
We are always then looking back at indicators on close to check if now meet and trigger buy/sell code. 

The first opportunity to know this is the open price of the current bar.

In the example, on entering the trade . We know when  00:00 to 03:59:59 closes the conditions are met. We can enter the trade at the "close 00-04am" price or the "open 04-08am" price. We would not wait till 07:59:59 for the "close 04-08am bar" value.

The same is true for closing trades. 

From analyzing the OHLC data around this recorded trade 
I would expect to see 0.28BTC profit not 0.31 as recorded. 

<img width="374" alt="screen shot 2018-06-27 at 10 09 28 am" src="https://user-images.githubusercontent.com/34645187/41967973-9184c2f8-79f2-11e8-8bf6-bad556729887.png">

<img width="269" alt="screen shot 2018-06-27 at 10 09 34 am" src="https://user-images.githubusercontent.com/34645187/41967982-9778b782-79f2-11e8-88cb-980f1a123cf7.png">

**Comment by creslinux** [comment]
I can confirm this is a bug. 
using the PR for plotting results to graph, this includes this paid and sold prices. 

Here we're using the close price of the 4am to 8am candle to buy 
based on triggers from the 0am to 4am candle. 
The price to use should have been open of 4am-8am or close of 0am to 8am. 

<img width="788" alt="screen shot 2018-06-27 at 11 26 42 am" src="https://user-images.githubusercontent.com/34645187/41971233-3b11fffc-79fd-11e8-99f4-888e857645b4.png">

**Comment by xmatthias** [comment]
Exporting of more fields will be added with #954

**Comment by creslinux** [comment]
The exporting of more fields is great, both as understanding what was paid and received is a core need. 

This bug is more worrisome as it misrepresents profitability of a strategy..
It *may (not tested) also prevent re-entry of the market artificially as the pair are in a trade that should be closed. 

In summary impacting result accuracy and *possibly volume of trades during backtest period.

**Comment by creslinux** [comment]
Is there any concern this may be in live also? 
Or will the buy/sell be triggered at 04:00:01 am ..as example and not 07:59:59/8am

**Comment by xmatthias** [comment]
in the code i linked above - you could try to keep the last sell_row for another round and put that in the backtest-result object ... 

that should then show it in the backtest-plot ... hopefully ... ;)

**Comment by creslinux** [comment]
@xmatthias  awesome - thanks for the tip
I'll give this a try today and test the results

**Comment by creslinux** [comment]
@xmatthias  patch fixes this issues.
https://github.com/xmatthias/freqtrade/tree/test/use_open_backtest
 
It has been tested and confirm the date/time recorded in trades table now have the price at that time in OHLC recorded. 

Before Patch trades table recorded a trade opening at 4am with the 8am price
<img width="1132" alt="screen shot 2018-07-05 at 6 57 41 pm" src="https://user-images.githubusercontent.com/34645187/42344891-381674c2-808d-11e8-8b2f-38e4bde885a4.png">

After Patch Trade entered at 4m has the 4am price.
<img width="996" alt="screen shot 2018-07-05 at 7 26 31 pm" src="https://user-images.githubusercontent.com/34645187/42344943-6604eda0-808d-11e8-969f-c212d0a2f87d.png">

**Comment by creslinux** [comment]
write up of patch results to be be clearer on the problem and fix. 
<img width="511" alt="screen shot 2018-07-06 at 7 50 01 am" src="https://user-images.githubusercontent.com/34645187/42368021-7917b1b2-80f5-11e8-8fba-13c616d1bb94.png">

**Comment by xmatthias** [comment]
i will submit a PR - but i'd like to first add some tests which show / confirm the behaviour as desired, and to avoid regressions in the future as this went undetected for ... long :)

**Comment by creslinux** [comment]
Seems a delay until tests are wrapping it would not hurt, it's been long
standing, does not impact live trading beyond strategy selection

**Comment by xmatthias** [comment]
beeing honest - #984 did not fix this but added a test for this - therefore reopening
